File: JsonSchemaAndStructs/KiteConnect-Dart-Structs/ohlc.dart
// To parse this JSON data, do
//
//     final ohlc = ohlcFromMap(jsonString);

import 'package:freezed_annotation/freezed_annotation.dart';
import 'dart:convert';

part 'ohlc.freezed.dart';
part 'ohlc.g.dart';

@freezed
abstract class Ohlc with _$Ohlc {
    const factory Ohlc({
        Map<String, Datum> data,
        String status,
    }) = _Ohlc;

    factory Ohlc.fromJson(Map<String, dynamic> json) => _$OhlcFromJson(json);
}

@freezed
abstract class Datum with _$Datum {
    const factory Datum({
        int instrumentToken,
        int lastPrice,
        OhlcClass ohlc,
    }) = _Datum;

    factory Datum.fromJson(Map<String, dynamic> json) => _$DatumFromJson(json);
}

@freezed
abstract class OhlcClass with _$OhlcClass {
    const factory OhlcClass({
        double close,
        double high,
        double low,
        double open,
    }) = _OhlcClass;

    factory OhlcClass.fromJson(Map<String, dynamic> json) => _$OhlcClassFromJson(json);
}

File: JsonSchemaAndStructs/KiteConnect-Haskell-Structs/ohlc.hs
{-# LANGUAGE StrictData #-}
{-# LANGUAGE OverloadedStrings #-}

module Ohlc
    ( Ohlc (..)
    , Datum (..)
    , OhlcClass (..)
    , decodeTopLevel
    ) where

import Data.Aeson
import Data.Aeson.Types (emptyObject)
import Data.ByteString.Lazy (ByteString)
import Data.HashMap.Strict (HashMap)
import Data.Text (Text)
import Data.Vector (Vector)

data Ohlc = Ohlc
    { ohlcDataOhlc :: Maybe (HashMap Text Datum)
    , statusOhlc :: Maybe Text
    } deriving (Show)

data Datum = Datum
    { instrumentTokenDatum :: Maybe Int
    , lastPriceDatum :: Maybe Int
    , ohlcDatum :: Maybe OhlcClass
    } deriving (Show)

data OhlcClass = OhlcClass
    { closeOhlcClass :: Maybe Float
    , highOhlcClass :: Maybe Float
    , lowOhlcClass :: Maybe Float
    , openOhlcClass :: Maybe Float
    } deriving (Show)

decodeTopLevel :: ByteString -> Maybe Ohlc
decodeTopLevel = decode

instance ToJSON Ohlc where
    toJSON (Ohlc ohlcDataOhlc statusOhlc) =
        object
        [ "data" .= ohlcDataOhlc
        , "status" .= statusOhlc
        ]

instance FromJSON Ohlc where
    parseJSON (Object v) = Ohlc
        <$> v .:? "data"
        <*> v .:? "status"

instance ToJSON Datum where
    toJSON (Datum instrumentTokenDatum lastPriceDatum ohlcDatum) =
        object
        [ "instrument_token" .= instrumentTokenDatum
        , "last_price" .= lastPriceDatum
        , "ohlc" .= ohlcDatum
        ]

instance FromJSON Datum where
    parseJSON (Object v) = Datum
        <$> v .:? "instrument_token"
        <*> v .:? "last_price"
        <*> v .:? "ohlc"

instance ToJSON OhlcClass where
    toJSON (OhlcClass closeOhlcClass highOhlcClass lowOhlcClass openOhlcClass) =
        object
        [ "close" .= closeOhlcClass
        , "high" .= highOhlcClass
        , "low" .= lowOhlcClass
        , "open" .= openOhlcClass
        ]

instance FromJSON OhlcClass where
    parseJSON (Object v) = OhlcClass
        <$> v .:? "close"
        <*> v .:? "high"
        <*> v .:? "low"
        <*> v .:? "open"
